lib/github.com/diku-dk/sorts/radix_sort.fut

```diff
@@ -109,3 +109,184 @@ def radix_sort_float_by_key [n] 't 'k
+  let offsets = scan (pairwise (+)) (0,0,0,0) flags
+  let (na,nb,nc,nd) = last offsets
+  let f bin (a,b,c,d) = i64.i16 ((-1)
+      + a * (i16.bool (bin == 0)) + na * (i16.bool (bin > 0))
+      + b * (i16.bool (bin == 1)) + nb * (i16.bool (bin > 1))
+      + c * (i16.bool (bin == 2)) + nc * (i16.bool (bin > 2))
+      + d * (i16.bool (bin == 3))) 
+  let is = map2 f bins offsets
+  in (scatter (copy xs) is xs, (na, nb, nc, nd))
+
+local def chunked_radix_sort_step [n] [m] [r] 't
+                          (get_bit: i32 -> t -> i32)
+                          (digit_n: i32)
+                          (xs: *[n*m+r]t) =
+  let hist_size = 4
+  let sort_step  arr =
+    let (ys, (a, b, c, d)) =
+      radix_sort_step_i16 arr get_bit digit_n
+    let hist' = sized hist_size [i64.i16 a
+                                ,i64.i16 b
+                                ,i64.i16 c
+                                ,i64.i16 d]
+    in (ys, hist')
+  let (chunks, rest) = split xs
+  let (sorted_rest, hist_rest) =
+    if r == 0
+    then ([], sized hist_size [0, 0, 0, 0])
+    else sort_step rest
+  let (sorted_chunks, hist_chunks) =
+    unflatten chunks
+    |> map sort_step
+    |> unzip
+  let histograms = hist_chunks ++ [hist_rest]
+  let ys = flatten sorted_chunks ++ sorted_rest
+  let flat_trans_hist =
+    histograms
+    |> transpose
+    |> flatten
+  let flat_hist =
+    histograms
+    |> flatten
+    |> sized (hist_size * (n + 1))
+  let (flat_trans_hist_scan, flat_hist_scan) =
+    zip flat_trans_hist flat_hist
+    |> exscan (\(a, b) (x, y) -> (a + x, b + y)) (0, 0)
+    |> unzip
+  let hist_scan =
+    flat_hist_scan
+    |> sized ((n + 1) * hist_size)
+    |> unflatten
+  let trans_hist_scan =
+    flat_trans_hist_scan
+    |> unflatten
+  let (is, elems) =
+    map (
+      \i ->
+        let elem = ys[i]
+        let bin = get_bin get_bit digit_n elem
+        let new_offset = trans_hist_scan[bin][i / m]
+        let old_offset = hist_scan[i / m][bin]
+        let idx = (i - old_offset) + new_offset
+        in (idx, elem)
+    ) (iota (n * m + r))
+    |> unzip
+  in scatter xs is elems
+
+local def (///) (a: i32) (b: i32) : i32 =
+  a / b + i32.bool (a % b != 0)
+
+local def (////) (a: i64) (b: i64) : i64 =
+  a / b + i64.bool (a % b != 0)
+
+-- | This implementation of radix sort works on the outside almost like
+-- `radix_sort` but the implementation is based of a design where you
+-- chunk the input into subarrays [1]. This leads to performance gains
+-- if you choose a good `chunk` size based on the GPU thread block.
+-- Using 512 as a `chunk` size leads to about twice the speed on
+-- a RTX 3060 as the normal `radix_sort`. The implementation also
+-- needs a `highest` element which is used for padding when sorting.
+-- The `highest` element needs to be larger or equal to the largest
+-- element in the input array. The sorting algorithm is stable and its
+-- work is *O(k n)* and the span is *O(k log(n))* as in `radix_sort`
```

Comment:
What is `k`?

Reply:
I now mention that this is the number of bits in the elements.